drivers/clock_control/clock_control_ra.c

```diff
@@ -0,0 +1,249 @@
+#define PLL_MULT(pll) DT_PROP(DT_INST_CHILD(0, pll), clock_mult)
+
+#define PLLMUL_BITS ((PLL_MULT(pll) << 1) - 1U)
+
+#if PLL_SRC_IS(mosc)
+#define PLSRCSEL_BIT (0)
+#elif PLL_SRC_IS(hoco)
+#define PLSRCSEL_BIT (1)
+#else
+#if IS_CLOCK_ENABLED(pll)
+#error Invalid PLL Source configuration
+#endif
+#endif
+
+#define PLLCCR_INIT_VALUE                                                                          \
+	(((PLLMUL_BITS << PLLCCR_PLLMUL_POS) | (PLSRCSEL_BIT << PLLCCR_PLSRCSEL_POS)) |            \
+	 (PLL_DIV(pll) << PLLCCR_PLLDIV_POS))
+
+#define SCKDIVCR_BITS(clk)                                                                         \
+	COND_CODE_1(IS_CLOCK_EXISTS(clk), ((SYSCLK_DIV(clk) & 0xFU) << SYSCLK_DIV_POS_##clk), (0U))
+
+#define SCKDIVCR_INIT_VALUE                                                                        \
+	(SCKDIVCR_BITS(iclk) | SCKDIVCR_BITS(pclka) | SCKDIVCR_BITS(pclkb) |                       \
+	 SCKDIVCR_BITS(pclkc) | SCKDIVCR_BITS(pclkd) | SCKDIVCR_BITS(bclk) | SCKDIVCR_BITS(fclk))
+
+#define HOCOWTCR_INIT_VALUE 6
+
+enum {
+	SYSCLK_DIV_1 = 0,
+	SYSCLK_DIV_2,
+	SYSCLK_DIV_4,
+	SYSCLK_DIV_8,
+	SYSCLK_DIV_16,
+	SYSCLK_DIV_32,
+	SYSCLK_DIV_64,
+	SYSCLK_DIV_128,
+	SYSCLK_DIV_3,
+	SYSCLK_DIV_6,
+	SYSCLK_DIV_12
+};
+
+enum {
+	PLL_DIV_1 = 0,
+	PLL_DIV_2,
+	PLL_DIV_3,
+	PLL_DIV_4,
+	PLL_DIV_5,
+	PLL_DIV_6,
+	PLL_DIV_8,
+	PLL_DIV_9,
+	PLL_DIV_16 = 15
+};
+
+enum {
+	SYSCLK_DIV_POS_pclkd = 0x0U,
+	SYSCLK_DIV_POS_pclkc = 0x4U,
+	SYSCLK_DIV_POS_pclkb = 0x8U,
+	SYSCLK_DIV_POS_pclka = 0xcU,
+	SYSCLK_DIV_POS_bclk = 0x10U,
+	SYSCLK_DIV_POS_pclke = 0x14U,
+	SYSCLK_DIV_POS_iclk = 0x18U,
+	SYSCLK_DIV_POS_fclk = 0x1cU
+};
+
+enum {
+	PLLCCR_PLLDIV_POS = 0,
+	PLLCCR_PLSRCSEL_POS = 4,
+	PLLCCR_PLLMUL_POS = 8
+};
+
+enum {
+	OSCSF_PLLSF_Msk = 0x20UL,
+	OSCSF_HOCOSF_Msk = 0x1UL,
+	SOPCCR_SOPCMTSF_Msk = 0x10UL
+};
+
+static const uint32_t PRCR_KEY = 0xA500U;
+static const uint32_t PRCR_CLOCKS = 0x1U;
+static const uint32_t PRCR_LOW_POWER = 0x2U;
+
+struct mstp_regs {
+#if DT_INST_REG_SIZE_BY_NAME(0, mstp) != 16
+	volatile uint32_t MSTPCRA;
+#endif
+	volatile uint32_t MSTPCRB;
+	volatile uint32_t MSTPCRC;
+	volatile uint32_t MSTPCRD;
+	volatile uint32_t MSTPCRE;
+};
+
+struct system_regs {
+	volatile uint8_t UNUSED0[32];
+	volatile uint32_t SCKDIVCR;
+	volatile uint8_t UNUSED1[2];
+	volatile uint8_t SCKSCR;
+	volatile uint8_t UNUSED2;
+	volatile uint16_t PLLCCR;
+	volatile uint8_t PLLCR;
+	volatile uint8_t PLLCCR2;
+	volatile uint8_t UNUSED3[5];
+	volatile uint8_t MEMWAIT;
+	volatile uint8_t MOSCCR;
+	volatile uint8_t UNUSED4[3];
+	volatile uint8_t HOCOCR;
+	volatile uint8_t UNUSED5[5];
+	volatile uint8_t OSCSF;
+	volatile uint8_t UNUSED6[99];
+	volatile uint8_t OPCCR;
+	volatile uint8_t UNUSED7[4];
+	volatile uint8_t HOCOWTCR;
+	volatile uint8_t UNUSED8[856];
+	volatile uint16_t PRCR;
+	volatile uint8_t UNUSED9[128];
+	volatile uint8_t SOSCCR;
+	volatile uint8_t UNUSED10[639];
+};
```

Comment:
suggestion: to improve code readability and consistency with native drivers, better define offsets and use the `sys_write/read` APIs.

Reply:
Fixed it.